pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -409,11 +424,133 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
         self._logger.warn("Issue %s is permitted with justification: %s", err, justification)
         return justification
 
-    async def publish_client(self, working_dir, from_release_tag, release_name, build_arch, client_type):
-        _, minor = util.isolate_major_minor_in_group(self.group)
-        quay_url = constants.QUAY_RELEASE_REPO_URL
+    async def extract_and_publish_clients(self, client_type: str, release_infos: Dict):
+        logger = self._logger
+        # make sure login to quay
+        if "QUAY_PASSWORD" in os.environ:
+            cmd = ["docker", "login", "-u", "openshift-release-dev+art_quay_dev", "-p", f"{os.environ['QUAY_PASSWORD']}", "quay.io"]
+            await exectools.cmd_assert_async(cmd, env=os.environ.copy(), stdout=sys.stderr)
+        base_to_mirror_dir = f"{self._working_dir}/to_mirror/openshift-v4"
+        message_digests = []
+        for arch, release_info in release_infos.items():
+            logger.info("Extracting client binaries for %s", arch)
+            pullspec = release_info["image"]
+            release_name = release_info["metadata"]["version"]
+            if arch == "multi":
+                manifest_arches = [brew_arch_for_go_arch(manifest["platform"]["architecture"]) for manifest in release_info.get("manifests", [])]
+                message_digest = await self.publish_multi_client(base_to_mirror_dir, pullspec, release_name, manifest_arches, client_type)
+            else:
+                message_digest = await self.publish_client(base_to_mirror_dir, pullspec, release_name, arch, client_type)
+            message_digests.append(message_digest)
+        return message_digests
+
+    async def sign_artifacts(self, release_name: str, client_type: str, release_infos: Dict, message_digests: List[str]):
+        """ Signs artifacts and publishes signature files to mirror
+        """
+        if not self.signing_env:
+            raise ValueError("--signing-env is missing")
+        cert_file = os.environ["SIGNING_CERT"]
+        key_file = os.environ["SIGNING_KEY"]
+        uri = constants.UMB_BROKERS[self.signing_env]
+        sig_keyname = "redhatrelease2" if client_type == 'ocp' else "beta2"
+        self._logger.info("About to sign artifacts with key %s", sig_keyname)
+        json_digest_sig_dir = self._working_dir / "json_digests"
+        message_digest_sig_dir = self._working_dir / "message_digests"
+        base_to_mirror_dir = self._working_dir / "to_mirror/openshift-v4"
+
+        lock_name = f'signing-lock'
+        lock_policy = locks.LOCK_POLICY['compose']
```

Comment:
Why not defining its own policy? Unless there's reasons to assimilate compose and signing locks, I'd rather keep them separate

Reply:
This is actually a copy-pasta XD. Will fix it in the next push.